Kernel/ACPI/ACPIStaticParser.cpp

```diff
@@ -117,39 +100,91 @@ void ACPIStaticParser::do_acpi_shutdown()
     ASSERT_NOT_REACHED();
 }
 
-void ACPIStaticParser::initialize_main_system_description_table()
+inline bool validate_acpi_table(ACPI_RAW::SDTHeader& v_header, size_t length)
+{
+    u8 checksum = 0;
+    auto* sdt = (u8*)&v_header;
+    for (size_t i = 0; i < length; i++)
+        checksum += sdt[i];
+    if (checksum == 0)
+        return true;
+    return false;
+}
+
+size_t ACPIStaticParser::get_table_size(ACPI_RAW::SDTHeader& p_header, const char* sig)
 {
-    auto checkup_region = MM.allocate_kernel_region((PAGE_SIZE * 2), "ACPI Static Parser Copying Method", Region::Access::Read);
+    InterruptDisabler disabler;
 #ifdef ACPI_DEBUG
-    dbgprintf("ACPI: Checking Main SDT Length to choose the correct mapping size");
+    dbgprintf("ACPI: Checking SDT Length\n");
 #endif
-    mmap_region(*checkup_region, PhysicalAddress((u32)m_main_system_description_table & PAGE_MASK));
+    auto region = MM.allocate_kernel_region(PhysicalAddress((u32)&p_header & PAGE_MASK), (PAGE_SIZE * 2), "ACPI get_table_size()", Region::Access::Read);
+    volatile auto* sdt = (ACPI_RAW::SDTHeader*)(create_vm_address(region->vaddr(), PhysicalAddress((u32)&p_header))).as_ptr();
+    // NOTE & FIXME: Apparently we need a loop to make sure we don't read some random garbage...
+    while (strncmp(const_cast<const char*>(sdt->sig), sig, strlen(sig)))
+        ;
+    ;
+    return sdt->length;
+}
 
-    auto* sdt = (ACPI_RAW::SDTHeader*)(checkup_region->vaddr().get() + ((u32)m_main_system_description_table & (~PAGE_MASK)));
-    u32 length = sdt->length;
-    u8 revision = sdt->revision;
+u8 ACPIStaticParser::get_table_revision(ACPI_RAW::SDTHeader& p_header, const char* sig)
+{
+    InterruptDisabler disabler;
+#ifdef ACPI_DEBUG
+    dbgprintf("ACPI: Checking SDT Revision\n");
+#endif
+    auto region = MM.allocate_kernel_region(PhysicalAddress((u32)&p_header & PAGE_MASK), (PAGE_SIZE * 2), "ACPI get_table_revision()", Region::Access::Read);
+    volatile auto* sdt = (ACPI_RAW::SDTHeader*)(create_vm_address(region->vaddr(), PhysicalAddress((u32)&p_header))).as_ptr();
+    // NOTE & FIXME: Apparently we need a loop to make sure we don't read some random garbage...
+    while (strncmp(const_cast<const char*>(sdt->sig), sig, strlen(sig)))
+        ;
```

Comment:
Same question here.

Reply:
We discussed it in the IRC channel. Apparently, we didn't flush the paging tables though I assumed we did (that was the problem for days that I couldn't figure what's going on!).
Fixed.